File: src/lib/hooks.ts
/**
 * Hooks 管理模块
 * 负责 skill hooks 与 .claude/settings.json 的合并和移除
 */
import { homedir } from 'os'
import { join, basename } from 'path'
import { existsSync, readFileSync, writeFileSync } from 'fs'
import { verbose } from './log.js'

/**
 * Claude Code settings.json 路径
 */
export function getClaudeSettingsPath(): string {
  return join(homedir(), '.claude', 'settings.json')
}

/**
 * Hook 配置项
 */
export interface HookConfig {
  type: 'command'
  command: string
  async?: boolean
  [key: string]: unknown
}

/**
 * Hook 匹配器配置
 */
export interface HookMatcher {
  matcher?: string | null
  hooks: HookConfig[]
}

/**
 * Hooks 配置结构
 */
export interface HooksConfig {
  [hookType: string]: HookMatcher[]
}

/**
 * Settings.json 结构
 */
export interface ClaudeSettings {
  hooks?: HooksConfig
  [key: string]: unknown
}

/**
 * Skill 的 hooks.json 结构
 */
export interface SkillHooksJson {
  hooks: HooksConfig
}

/**
 * 获取 skill 标识符（用于 command 注释）
 */
function getSkillMarker(skillName: string): string {
  return `# skill: ${skillName}`
}

/**
 * 检查 command 是否来自指定 skill
 */
function isCommandFromSkill(command: string, skillName: string): boolean {
  return command.includes(getSkillMarker(skillName))
}

/**
 * 读取 settings.json
 */
export function readClaudeSettings(): ClaudeSettings {
  const settingsPath = getClaudeSettingsPath()

  if (!existsSync(settingsPath)) {
    return {}
  }

  try {
    const content = readFileSync(settingsPath, 'utf-8')
    return JSON.parse(content) as ClaudeSettings
  } catch (error) {
    verbose(`Failed to read settings.json: ${(error as Error).message}`)
    return {}
  }
}

/**
 * 写入 settings.json
 */
export function writeClaudeSettings(settings: ClaudeSettings): boolean {
  const settingsPath = getClaudeSettingsPath()

  try {
    writeFileSync(settingsPath, JSON.stringify(settings, null, 2), 'utf-8')
    return true
  } catch (error) {
    verbose(`Failed to write settings.json: ${(error as Error).message}`)
    return false
  }
}

/**
 * 读取 skill 的 hooks.json
 */
export function readSkillHooks(skillPath: string): SkillHooksJson | null {
  const hooksPath = join(skillPath, 'hooks', 'hooks.json')

  if (!existsSync(hooksPath)) {
    return null
  }

  try {
    const content = readFileSync(hooksPath, 'utf-8')
    return JSON.parse(content) as SkillHooksJson
  } catch (error) {
    verbose(`Failed to read skill hooks.json: ${(error as Error).message}`)
    return null
  }
}

/**
 * 检查 skill 是否有 hooks
 */
export function hasSkillHooks(skillPath: string): boolean {
  return readSkillHooks(skillPath) !== null
}

/**
 * 替换 hooks.json 中的变量
 * ${CLAUDE_PLUGIN_ROOT} -> skill 实际路径
 */
function resolveHookVariables(command: string, skillPath: string): string {
  return command.replace(/\$\{CLAUDE_PLUGIN_ROOT\}/g, skillPath)
}

/**
 * 验证 hooks 配置结构是否符合规范
 * 在合并/移除操作前调用，防止损坏 settings.json
 */
export function validateHooksConfig(hooks: unknown): { valid: boolean; errors: string[] } {
  const errors: string[] = []

  // 1. hooks 必须是非 null 的 object，且不是数组
  if (hooks === null || hooks === undefined || typeof hooks !== 'object' || Array.isArray(hooks)) {
    return { valid: false, errors: ['hooks 必须是非 null 的 object'] }
  }

  const hooksObj = hooks as Record<string, unknown>

  for (const [hookType, matchers] of Object.entries(hooksObj)) {
    // 2. 每个 hookType 的值必须是数组
    if (!Array.isArray(matchers)) {
      errors.push(`hooks.${hookType} 不是数组`)
      continue
    }

    matchers.forEach((matcher, mIdx) => {
      // 3. 每个 matcher 必须是 object
      if (matcher === null || matcher === undefined || typeof matcher !== 'object' || Array.isArray(matcher)) {
        errors.push(`hooks.${hookType}[${mIdx}] 不是 object`)
        return
      }

      const m = matcher as Record<string, unknown>

      // 4. matcher 字段如果存在必须是 string 或 null
      if ('matcher' in m && m.matcher !== null && typeof m.matcher !== 'string') {
        errors.push(`hooks.${hookType}[${mIdx}].matcher 不是 string/null`)
      }

      // 5. 必须有 hooks 属性且为数组
      if (!('hooks' in m)) {
        errors.push(`hooks.${hookType}[${mIdx}] 缺少 hooks 属性`)
        return
      }
      if (!Array.isArray(m.hooks)) {
        errors.push(`hooks.${hookType}[${mIdx}].hooks 不是数组`)
        return
      }

      // 6. 每个 hook 必须有 type(string) 和 command(string)
      m.hooks.forEach((hook: unknown, hIdx: number) => {
        if (hook === null || hook === undefined || typeof hook !== 'object' || Array.isArray(hook)) {
          errors.push(`hooks.${hookType}[${mIdx}].hooks[${hIdx}] 不是 object`)
          return
        }
        const h = hook as Record<string, unknown>
        if (!('type' in h) || typeof h.type !== 'string') {
          errors.push(`hooks.${hookType}[${mIdx}].hooks[${hIdx}] 缺少 type 或 type 不是 string`)
        }
        if (!('command' in h) || typeof h.command !== 'string') {
          errors.push(`hooks.${hookType}[${mIdx}].hooks[${hIdx}] 缺少 command 或 command 不是 string`)
        }
      })
    })
  }

  return { valid: errors.length === 0, errors }
}

/**
 * 归一化 matcher：null/undefined 视为空 matcher
 */
function normalizeMatcher(matcher: string | null | undefined): string {
  return matcher ?? ''
}

/**
 * 合并 skill hooks 到 settings.json
 * @param skillPath skill 路径
 * @param skillName skill 名称
 * @returns 是否成功合并
 */
export function mergeSkillHooks(skillPath: string, skillName: string): boolean {
  const skillHooks = readSkillHooks(skillPath)

  if (!skillHooks || !skillHooks.hooks) {
    verbose(`No hooks found in skill: ${skillName}`)
    return false
  }

  // 前置验证：检查 skill hooks.json 结构
  const skillValidation = validateHooksConfig(skillHooks.hooks)
  if (!skillValidation.valid) {
    verbose(`skill hooks.json 结构异常，中止合并: ${skillValidation.errors.join('; ')}`)
    return false
  }

  const settings = readClaudeSettings()

  // 前置验证：检查 settings.json 现有 hooks 结构
  if (settings.hooks) {
    const validation = validateHooksConfig(settings.hooks)
    if (!validation.valid) {
      verbose(`settings.json hooks 结构异常，中止合并: ${validation.errors.join('; ')}`)
      return false
    }
  }

  // 确保 hooks 对象存在
  if (!settings.hooks) {
    settings.hooks = {}
  }

  const skillMarker = getSkillMarker(skillName)

  // 遍历 skill 的所有 hook 类型
  for (const [hookType, matchers] of Object.entries(skillHooks.hooks)) {
    // 确保该 hook 类型存在
    if (!settings.hooks[hookType]) {
      settings.hooks[hookType] = []
    }

    // 遍历每个 matcher 配置
    for (const matcher of matchers) {
      // 转换 hooks 中的 command
      const resolvedHooks: HookConfig[] = matcher.hooks.map((hook) => {
        const resolved: HookConfig = {
          type: hook.type,
          command: `${resolveHookVariables(hook.command, skillPath)} ${skillMarker}`,
        }
        // 保留 async 等额外字段
        for (const [key, value] of Object.entries(hook)) {
          if (key !== 'type' && key !== 'command') {
            resolved[key] = value
          }
        }
        return resolved
      })

      // 查找是否已有相同 matcher 的配置
      const incomingMatcher = normalizeMatcher(matcher.matcher)
      const existingMatcher = settings.hooks[hookType].find(
        (m) => normalizeMatcher(m.matcher) === incomingMatcher
      )

      if (existingMatcher) {
        // 统一 matcher 表达，避免 null/undefined 与空字符串混用
        existingMatcher.matcher = incomingMatcher

        // 合并 hooks（避免重复）
        for (const resolvedHook of resolvedHooks) {
          const exists = existingMatcher.hooks.some(
            (h) => h.command === resolvedHook.command
          )
          if (!exists) {
            existingMatcher.hooks.push(resolvedHook)
          }
        }
      } else {
        // 添加新的 matcher 配置
        settings.hooks[hookType].push({
          matcher: incomingMatcher,
          hooks: resolvedHooks,
        })
      }
    }

    verbose(`Merged hooks for ${hookType} in skill: ${skillName}`)
  }

  return writeClaudeSettings(settings)
}

/**
 * 从 settings.json 移除 skill hooks
 * @param skillName skill 名称
 * @returns 是否成功移除
 */
export function removeSkillHooks(skillName: string): boolean {
  const settings = readClaudeSettings()

  if (!settings.hooks) {
    verbose('No hooks found in settings.json')
    return false
  }

  // 前置验证：检查 hooks 结构
  const validation = validateHooksConfig(settings.hooks)
  if (!validation.valid) {
    verbose(`settings.json hooks 结构异常，中止移除: ${validation.errors.join('; ')}`)
    return false
  }

  const skillMarker = getSkillMarker(skillName)
  let removed = false

  // 遍历所有 hook 类型
  for (const [hookType, matchers] of Object.entries(settings.hooks)) {
    // 遍历每个 matcher 配置
    for (let i = matchers.length - 1; i >= 0; i--) {
      const matcher = matchers[i]

      // 过滤掉来自该 skill 的 hooks
      const filteredHooks = matcher.hooks.filter(
        (hook) => !isCommandFromSkill(hook.command, skillName)
      )

      if (filteredHooks.length === 0) {
        // 如果没有剩余的 hooks，移除整个 matcher
        matchers.splice(i, 1)
        removed = true
        verbose(`Removed empty matcher for ${hookType}`)
      } else if (filteredHooks.length !== matcher.hooks.length) {
        // 更新 hooks 列表
        matcher.hooks = filteredHooks
        removed = true
        verbose(`Removed hooks from ${hookType} for skill: ${skillName}`)
      }
    }

    // 如果该 hook 类型没有 matchers 了，移除整个 hook 类型
    if (matchers.length === 0) {
      delete settings.hooks[hookType]
      verbose(`Removed empty hook type: ${hookType}`)
    }
  }

  if (removed) {
    return writeClaudeSettings(settings)
  }

  return false
}

/**
 * 检查 settings.json 中是否包含指定 skill 的 hooks
 */
export function hasSkillHooksInSettings(skillName: string): boolean {
  const settings = readClaudeSettings()

  if (!settings.hooks) {
    return false
  }

  for (const matchers of Object.values(settings.hooks)) {
    for (const matcher of matchers) {
      for (const hook of matcher.hooks) {
        if (isCommandFromSkill(hook.command, skillName)) {
          return true
        }
      }
    }
  }

  return false
}

/**
 * 列出 settings.json 中所有 skill hooks
 * @returns skill 名称列表
 */
export function listInstalledSkillHooks(): string[] {
  const settings = readClaudeSettings()
  const skills = new Set<string>()

  if (!settings.hooks) {
    return []
  }

  const markerRegex = /# skill: ([^\s]+)/

  for (const matchers of Object.values(settings.hooks)) {
    for (const matcher of matchers) {
      for (const hook of matcher.hooks) {
        const match = hook.command.match(markerRegex)
        if (match) {
          skills.add(match[1])
        }
      }
    }
  }

  return Array.from(skills)
}

File: src-tauri/src/services/hooks.rs
//! Hooks 管理模块
//! 负责 skill hooks 与 .claude/settings.json 的合并和移除

use crate::utils::paths::get_claude_settings_path;
use crate::Result;
use serde::de::Deserializer;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;
use std::fs;
use std::path::Path;

/// Hook 配置项
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HookConfig {
    #[serde(rename = "type")]
    pub hook_type: String,
    pub command: String,
}

/// Hook 匹配器配置
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HookMatcher {
    #[serde(default, deserialize_with = "deserialize_matcher")]
    pub matcher: String,
    pub hooks: Vec<HookConfig>,
}

/// 兼容历史配置：matcher 缺失/null 时归一化为空字符串
fn deserialize_matcher<'de, D>(deserializer: D) -> std::result::Result<String, D::Error>
where
    D: Deserializer<'de>,
{
    let value: Option<Option<String>> = Option::deserialize(deserializer)?;
    Ok(value.flatten().unwrap_or_default())
}

/// Hooks 配置结构（按 hook 类型分组）
pub type HooksConfig = HashMap<String, Vec<HookMatcher>>;

/// Skill 的 hooks.json 结构
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SkillHooksJson {
    pub hooks: HooksConfig,
}

/// 获取 skill 标识符（用于 command 注释）
fn get_skill_marker(skill_name: &str) -> String {
    format!("# skill: {}", skill_name)
}

/// 检查 command 是否来自指定 skill
fn is_command_from_skill(command: &str, skill_name: &str) -> bool {
    command.contains(&get_skill_marker(skill_name))
}

/// 读取 settings.json 为通用 JSON Value（保留所有字段）
fn read_claude_settings_json() -> Result<Value> {
    let settings_path = get_claude_settings_path()?;

    if !settings_path.exists() {
        return Ok(serde_json::json!({}));
    }

    let content = fs::read_to_string(&settings_path)?;
    let settings: Value = serde_json::from_str(&content)?;
    Ok(settings)
}

/// 写入 settings.json（保留所有字段）
fn write_claude_settings_json(settings: &Value) -> Result<()> {
    let settings_path = get_claude_settings_path()?;

    // 确保目录存在
    if let Some(parent) = settings_path.parent() {
        if !parent.exists() {
            fs::create_dir_all(parent)?;
        }
    }

    let content = serde_json::to_string_pretty(settings)?;
    fs::write(&settings_path, content)?;
    Ok(())
}

/// 从 JSON Value 中获取 hooks 配置
fn get_hooks_from_value(value: &Value) -> Option<HooksConfig> {
    value.get("hooks").and_then(|h| serde_json::from_value(h.clone()).ok())
}

/// 设置 JSON Value 中的 hooks 配置
fn set_hooks_in_value(value: &mut Value, hooks: &HooksConfig) {
    if let Ok(hooks_value) = serde_json::to_value(hooks) {
        value["hooks"] = hooks_value;
    }
}

/// 读取 skill 的 hooks.json
pub fn read_skill_hooks(skill_path: &Path) -> Result<Option<SkillHooksJson>> {
    let hooks_path = skill_path.join("hooks").join("hooks.json");

    if !hooks_path.exists() {
        return Ok(None);
    }

    let content = fs::read_to_string(&hooks_path)?;
    let hooks: SkillHooksJson = serde_json::from_str(&content)?;
    Ok(Some(hooks))
}

/// 检查 skill 是否有 hooks
pub fn has_skill_hooks(skill_path: &Path) -> bool {
    read_skill_hooks(skill_path).map_or(false, |h| h.is_some())
}

/// 验证 hooks 配置结构是否符合规范
/// 在合并/移除操作前调用，防止损坏 settings.json
pub fn validate_hooks_config(hooks: &serde_json::Value) -> std::result::Result<(), Vec<String>> {
    let mut errors = Vec::new();

    // 1. hooks 必须是 object
    if !hooks.is_object() {
        return Err(vec!["hooks 必须是 object".to_string()]);
    }

    if let Some(hooks_obj) = hooks.as_object() {
        for (hook_type, matchers_val) in hooks_obj {
            // 2. 每个 hookType 的值必须是数组
            if !matchers_val.is_array() {
                errors.push(format!("hooks.{} 不是数组", hook_type));
                continue;
            }

            if let Some(matchers) = matchers_val.as_array() {
                for (m_idx, matcher_val) in matchers.iter().enumerate() {
                    // 3. 每个 matcher 必须是 object
                    if !matcher_val.is_object() {
                        errors.push(format!("hooks.{}[{}] 不是 object", hook_type, m_idx));
                        continue;
                    }

                    let m = matcher_val.as_object().unwrap();

                    // 4. matcher 字段如果存在必须是 string 或 null（null 在读取时会归一化为 ""）
                    if let Some(mv) = m.get("matcher") {
                        if !mv.is_string() && !mv.is_null() {
                            errors.push(format!(
                                "hooks.{}[{}].matcher 不是 string/null",
                                hook_type, m_idx
                            ));
                        }
                    }

                    // 5. 必须有 hooks 属性且为数组
                    match m.get("hooks") {
                        None => {
                            errors.push(format!(
                                "hooks.{}[{}] 缺少 hooks 属性",
                                hook_type, m_idx
                            ));
                        }
                        Some(h) if !h.is_array() => {
                            errors.push(format!(
                                "hooks.{}[{}].hooks 不是数组",
                                hook_type, m_idx
                            ));
                        }
                        Some(hooks_arr) => {
                            // 6. 每个 hook 必须有 type(string) 和 command(string)
                            if let Some(arr) = hooks_arr.as_array() {
                                for (h_idx, hook_val) in arr.iter().enumerate() {
                                    if !hook_val.is_object() {
                                        errors.push(format!(
                                            "hooks.{}[{}].hooks[{}] 不是 object",
                                            hook_type, m_idx, h_idx
                                        ));
                                        continue;
                                    }
                                    let h = hook_val.as_object().unwrap();
                                    match h.get("type") {
                                        None => errors.push(format!(
                                            "hooks.{}[{}].hooks[{}] 缺少 type",
                                            hook_type, m_idx, h_idx
                                        )),
                                        Some(t) if !t.is_string() => errors.push(format!(
                                            "hooks.{}[{}].hooks[{}].type 不是 string",
                                            hook_type, m_idx, h_idx
                                        )),
                                        _ => {}
                                    }
                                    match h.get("command") {
                                        None => errors.push(format!(
                                            "hooks.{}[{}].hooks[{}] 缺少 command",
                                            hook_type, m_idx, h_idx
                                        )),
                                        Some(c) if !c.is_string() => errors.push(format!(
                                            "hooks.{}[{}].hooks[{}].command 不是 string",
                                            hook_type, m_idx, h_idx
                                        )),
                                        _ => {}
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }

    if errors.is_empty() {
        Ok(())
    } else {
        Err(errors)
    }
}

/// 替换 hooks.json 中的变量
/// ${CLAUDE_PLUGIN_ROOT} -> skill 实际路径
fn resolve_hook_variables(command: &str, skill_path: &Path) -> String {
    let skill_path_str = skill_path.to_string_lossy();
    command.replace("${CLAUDE_PLUGIN_ROOT}", &skill_path_str)
}

/// 合并 skill hooks 到 settings.json
pub fn merge_skill_hooks(skill_path: &Path, skill_name: &str) -> Result<bool> {
    let skill_hooks = match read_skill_hooks(skill_path)? {
        Some(h) => h,
        None => return Ok(false),
    };

    // 读取完整的 settings.json（保留所有字段）
    let mut settings = read_claude_settings_json()?;

    // 前置验证：检查 skill hooks.json 结构
    if let Ok(skill_hooks_value) = serde_json::to_value(&skill_hooks.hooks) {
        if let Err(e) = validate_hooks_config(&skill_hooks_value) {
            eprintln!("skill hooks.json 结构异常，中止合并: {}", e.join("; "));
            return Ok(false);
        }
    }

    // 前置验证：检查 settings.json 现有 hooks 结构
    if let Some(existing_hooks) = settings.get("hooks") {
        if let Err(e) = validate_hooks_config(existing_hooks) {
            eprintln!("settings.json hooks 结构异常，中止合并: {}", e.join("; "));
            return Ok(false);
        }
    }

    // 获取或创建 hooks 配置
    let mut hooks: HooksConfig = get_hooks_from_value(&settings).unwrap_or_default();

    let skill_marker = get_skill_marker(skill_name);

    // 遍历 skill 的所有 hook 类型
    for (hook_type, matchers) in skill_hooks.hooks {
        // 确保该 hook 类型存在
        let entry = hooks.entry(hook_type.clone()).or_insert_with(Vec::new);

        // 遍历每个 matcher 配置
        for matcher in matchers {
            // 转换 hooks 中的 command
            let resolved_hooks: Vec<HookConfig> = matcher
                .hooks
                .iter()
                .map(|hook| HookConfig {
                    hook_type: hook.hook_type.clone(),
                    command: format!(
                        "{} {}",
                        resolve_hook_variables(&hook.command, skill_path),
                        skill_marker
                    ),
                })
                .collect();

            // 查找是否已有相同 matcher 的配置
            let matcher_value = matcher.matcher.as_str();
            let existing_matcher = entry
                .iter_mut()
                .find(|m| m.matcher == matcher_value);

            if let Some(existing) = existing_matcher {
                // 合并 hooks（避免重复）
                for resolved_hook in resolved_hooks {
                    let exists = existing
                        .hooks
                        .iter()
                        .any(|h| h.command == resolved_hook.command);
                    if !exists {
                        existing.hooks.push(resolved_hook);
                    }
                }
            } else {
                // 添加新的 matcher 配置
                entry.push(HookMatcher {
                    matcher: matcher.matcher,
                    hooks: resolved_hooks,
                });
            }
        }
    }

    // 更新 settings 中的 hooks（保留其他字段）
    set_hooks_in_value(&mut settings, &hooks);
    write_claude_settings_json(&settings)?;
    Ok(true)
}

/// 从 settings.json 移除 skill hooks
pub fn remove_skill_hooks(skill_name: &str) -> Result<bool> {
    // 读取完整的 settings.json（保留所有字段）
    let mut settings = read_claude_settings_json()?;

    // 前置验证：检查 hooks 结构
    if let Some(existing_hooks) = settings.get("hooks") {
        if let Err(e) = validate_hooks_config(existing_hooks) {
            eprintln!("settings.json hooks 结构异常，中止移除: {}", e.join("; "));
            return Ok(false);
        }
    }

    let mut hooks = match get_hooks_from_value(&settings) {
        Some(h) => h,
        None => return Ok(false),
    };

    let mut removed = false;

    // 遍历所有 hook 类型
    let hook_types: Vec<String> = hooks.keys().cloned().collect();
    for hook_type in hook_types {
        if let Some(matchers) = hooks.get_mut(&hook_type) {
            // 遍历每个 matcher 配置
            let mut i = matchers.len();
            while i > 0 {
                i -= 1;
                let matcher = &mut matchers[i];

                // 过滤掉来自该 skill 的 hooks
                matcher.hooks.retain(|hook| !is_command_from_skill(&hook.command, skill_name));

                if matcher.hooks.is_empty() {
                    // 如果没有剩余的 hooks，移除整个 matcher
                    matchers.remove(i);
                    removed = true;
                } else if matcher.hooks.len() != matcher.hooks.len() {
                    removed = true;
                }
            }

            // 如果该 hook 类型没有 matchers 了，移除整个 hook 类型
            if matchers.is_empty() {
                hooks.remove(&hook_type);
            }
        }
    }

    if removed {
        set_hooks_in_value(&mut settings, &hooks);
        write_claude_settings_json(&settings)?;
    }

    Ok(removed)
}

/// 检查 settings.json 中是否包含指定 skill 的 hooks
pub fn has_skill_hooks_in_settings(skill_name: &str) -> bool {
    let settings = match read_claude_settings_json() {
        Ok(s) => s,
        Err(_) => return false,
    };

    if let Some(hooks) = get_hooks_from_value(&settings) {
        for matchers in hooks.values() {
            for matcher in matchers {
                for hook in &matcher.hooks {
                    if is_command_from_skill(&hook.command, skill_name) {
                        return true;
                    }
                }
            }
        }
    }

    false
}

/// 列出 settings.json 中所有 skill hooks
pub fn list_installed_skill_hooks() -> Result<Vec<String>> {
    let settings = read_claude_settings_json()?;
    let mut skills = std::collections::HashSet::new();

    if let Some(hooks) = get_hooks_from_value(&settings) {
        let marker_regex = regex::Regex::new(r"# skill: (\S+)").unwrap();

        for matchers in hooks.values() {
            for matcher in matchers {
                for hook in &matcher.hooks {
                    if let Some(caps) = marker_regex.captures(&hook.command) {
                        if let Some(name) = caps.get(1) {
                            skills.insert(name.as_str().to_string());
                        }
                    }
                }
            }
        }
    }

    Ok(skills.into_iter().collect())
}
